google/cloud/bigquery/client.py

```diff
@@ -1934,7 +1944,7 @@ def create_job(
             return self.load_table_from_uri(
                 source_uris,
                 destination,
-                job_config=load_job_config,
+                job_config=typing.cast(LoadJobConfig, load_job_config),
```

Comment:
The problem is that `from_api_repr()` is defined on the base `_JobConfig` class and its return type is annotated as `_JobConfig`. Which makes sense on the base class, but in practice the actual `cls` subclass is chosen here dynamically by introspecting `job_config`, and `mypy` cannot understand that.

Specifying the type of `load_job_config` at assignment time would not help, because again a base type (`_JobConfig`) is assigned to `load_job_config`  with a specific sub-type, and `mypy` would still complain. Cast solves this.

The alternative could be changing the return type on the base class to include all sub-classes, but that doesn't feel right. It can get of of sync if a new sub-type is created, and a circular import would be needed.
Besides, `mypy` would probably _still_ complain if `"Union[All, Config, Subtypes]"` is assigned to a _specific_ sub-class type.